Jenkinsfile

```diff
@@ -1,28 +1,110 @@
 #!groovy
 library "atomSharedLibraries@master"
 
+import static Constants.*
+
+class Constants {
+    static final mainDirectoryName = "contrail_state_files"
+    static final terraformStateFileName = "state"
+}
+
+def warningEcho(message) {
+    echo "\033[1;33m${message}\033[0m"
+}
+
+def ensureMainDirectoryExists() {
+    dirExistingResult = checkIfDirectoryExists("../${mainDirectoryName}")
+    if(dirExistingResult == '0') {
+        sh "mkdir ../${mainDirectoryName}"
+    }
+}
+
+def ensureUserDirectoryExists() {
+    dirExistingResult = checkIfDirectoryExists("../${mainDirectoryName}/${params.Login}")
+    if(dirExistingResult == '0') {
+        sh "mkdir ../${mainDirectoryName}/${params.Login}"
+    }
+}
+
+def checkIfDirectoryExists(name) {
+    return sh(script: "test -d \"${name}\" && echo '1' || echo '0' ", returnStdout: true).trim()
+}
+
+def checkIfFileExists(name) {
+    return sh(script: "test -f \"${name}\" && echo '1' || echo '0' ", returnStdout: true).trim()
+}
+
+def copyTerraformFiles() {
+    sh "cp provision/${params.orchestrator}/${params.orchestrator}.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/${params.orchestrator}.tf"
+    sh "cp provision/${params.orchestrator}/variables.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/variables.tf"
+}
+
+def copyDaemonFile() {
+    sh "cp provision/daemon.json ../${mainDirectoryName}/${params.Login}/${params.MachineName}/daemon.json"
+}
+
+def resolveInstancesYamlFile() {
+    if ("${instances_yaml}" != "") {
+        def instancesYaml = unstashParam "instances_yaml"
+        sh "mv ${instancesYaml} ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
+    } else {
+        warningEcho("Didn't get instances yaml file. Using default file.")
+        sh "cp provision/template.yaml ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
+    }
+}
+
+def prepareMachineDirectory() {
+    sh "mkdir ../${mainDirectoryName}/${params.Login}/${params.MachineName}"
+    sh "chmod 777 provision/prepare_template"
+    sh "cp provision/prepare_template ../${mainDirectoryName}/${params.Login}/${params.MachineName}/prepare_template"
+    copyTerraformFiles()
+    copyDaemonFile()
+    resolveInstancesYamlFile()
+    prepareKeyFiles()
+}
+
+def prepareKeyFiles() {
+    // This is a workaround!
+    // https://bitbucket.org/janvrany/jenkins-27413-workaround-library
+    if ("${sshpubkey}" != "" && "${sshprivkey}" != "") {
+        def sshPubKeyFile = unstashParam "sshpubkey"
+        def sshPrivKeyFile = unstashParam "sshprivkey"
+        sh "mv ${sshPubKeyFile} ../${mainDirectoryName}/${params.Login}/${params.MachineName}/${params.Login}-key.pub"
```

Comment:
`${mainDirectoryName}/${params.Login}/${params.MachineName}/` this path is used quite often. Maybe it could be returned by some function.

Reply:
Done.